edgelet/edgelet-http-workload/src/server/cert/mod.rs

```diff
@@ -62,34 +67,178 @@ fn compute_validity(expiration: &str, max_duration_sec: i64, context: ErrorKind)
     Ok(cmp::min(secs, max_duration_sec))
 }
 
-fn refresh_cert<T: CreateCertificate>(
-    hsm: &T,
+fn refresh_cert(
+    key_client: Arc<Mutex<aziot_key_client::Client>>,
+    cert_client: Arc<Mutex<CertificateClient>>,
     alias: String,
     props: &CertificateProperties,
     context: ErrorKind,
-) -> Result<Response<Body>> {
-    if let Err(err) = hsm.destroy_certificate(alias) {
-        return Err(Error::from(err.context(context)));
-    };
+) -> Box<dyn Future<Item = Response<Body>, Error = HttpError> + Send> {
+    let response = generate_key_and_csr(props, context.clone())
+        .map_err(|_| Error::from(ErrorKind::CertOperation(CertOperation::CreateIdentityCert)))
+        .and_then(|(privkey, csr)| {
+            let workload_ca_key_pair_handle = key_client.lock().expect("key client lock error").create_key_pair_if_not_exists(IOTEDGED_CA_ALIAS, Some("ec-p256:rsa-4096:*"))
+                .map_err(|_| Error::from(ErrorKind::CertOperation(CertOperation::CreateIdentityCert)))?;
+                Ok((privkey, csr, workload_ca_key_pair_handle))
+        })
+    .into_future()
+    .and_then(move |(privkey, csr, workload_ca_key_pair_handle)| -> Result<_> {
+        let response = cert_client
+            .lock()
+            .expect("certificate client lock error")
+            .create_cert(&alias, &csr, Some((IOTEDGED_CA_ALIAS, &workload_ca_key_pair_handle)))
+            .map_err(|e| Error::from(e.context(ErrorKind::CertOperation(CertOperation::CreateIdentityCert))))
+            .map(|cert| (privkey, cert))
+            .and_then( |(privkey, cert)| {
+                let pk = privkey.private_key_to_pem_pkcs8().context(context.clone())?;
+                let cert = Certificate::new(cert, pk);
+                let cert = cert_to_response(&cert, context.clone())?;
+                let body = match serde_json::to_string(&cert) {
+                    Ok(body) => body,
+                    Err(err) => return Err(Error::from(err.context(context))),
+                };
+                
+                let response = Response::builder()
+                .status(StatusCode::CREATED)
+                .header(CONTENT_TYPE, "application/json")
+                .header(CONTENT_LENGTH, body.len().to_string().as_str())
+                .body(body.into())
+                .context(context)?;
+                
+                Ok(response)
+            });
+        Ok(response)
+    })
+    .flatten()
+    .or_else(|e| Ok(e.into_response()));
+
+    Box::new(response)
+}
 
-    let cert = match hsm.create_certificate(props) {
-        Ok(cert) => cert,
-        Err(err) => return Err(Error::from(err.context(context))),
+fn generate_key_and_csr(
+    props: &CertificateProperties,
+    context: ErrorKind,
+) -> Result<(openssl::pkey::PKey<openssl::pkey::Private>, Vec<u8>)> {
+    let rsa = openssl::rsa::Rsa::generate(2048)
+        .context(context.clone())?;
```

Comment:
In general, `.context(expr_with_side_effects)` should be `.with_context(|_| expr_with_side_effects)`

In this case specifically, all your errors are `openssl::Error`, so you might as well just return that with `?` and do the context mapping in the single caller of this function.

Reply:
Changed to `openssl::Error`